ParProcCo/job_scheduler.py

```diff
@@ -42,50 +118,163 @@ def __init__(
+            user_token if user_token else os.environ["SLURM_JWT"]
+        )
 
-    def check_queue_list(self, queue: str) -> str:
-        if not queue:
-            raise ValueError("queue must be non-empty string")
-        queue = queue.lower()
-        with os.popen("qconf -sql") as q_proc:
-            q_name_list = q_proc.read().split()
-        if queue in q_name_list:
-            return queue
-        else:
-            raise ValueError(f"queue {queue} not in queue list: {q_name_list}\n")
-
-    def check_project_list(self, project: str) -> str:
-        if not project:
-            raise ValueError("project must be non-empty string")
-        with os.popen("qconf -sprjl") as prj_proc:
-            prj_name_list = prj_proc.read().split()
-        if project in prj_name_list:
-            return project
+    def get(
+        self,
+        endpoint: str,
+        params: dict[str, Any] | None = None,
+        timeout: float | None = None,
+    ) -> requests.Response:
+        response = self._session.get(
+            f"{self._url}/{endpoint}", params=params, timeout=timeout
+        )
+        response.raise_for_status()
+        return response
+
+    def put(
+        self,
+        endpoint: str,
+        params: dict[str, Any] | None = None,
+        json: dict[str, Any] | None = None,
+        timeout: float | None = None,
+    ) -> requests.Response:
+        response = self._session.put(
+            f"{self._url}/{endpoint}", params=params, json=json, timeout=timeout
+        )
+        response.raise_for_status()
+        return response
+
+    def _prepare_request(
+        self, data: BaseModel
+    ) -> tuple[str, dict[str, str]] | tuple[None, None]:
+        if data is None:
+            return None, None
+        return data.model_dump_json(exclude_defaults=True), {
+            "X-SLURM-USER-NAME": self.user,
+            "X-SLURM-USER-TOKEN": self.token,
+            "Content-Type": "application/json",
+        }
+
+    def _post(self, data: BaseModel, endpoint):
+        url = self._url + "/" + endpoint
+        jdata, headers = self._prepare_request(data)
+        resp = requests.post(url, data=jdata, headers=headers)
+        return resp
+
+    def delete(
+        self,
+        endpoint: str,
+        params: dict[str, Any] | None = None,
+        timeout: float | None = None,
+    ) -> requests.Response:
+        response = self._session.delete(
+            f"{self._url}/{endpoint}", params=params, timeout=timeout
+        )
+        response.raise_for_status()
+        return response
+
+    def get_jobs(self) -> JobsResponse:
+        endpoint = f"slurm/{self._version}/jobs"
+        response = self.get(endpoint)
+        return JobsResponse.model_validate(response.json())
+
+    def get_job_response(self, job_id: int) -> JobsResponse:
+        endpoint = f"slurm/{self._version}/job/{job_id}"
+        response = self.get(endpoint)
+        return JobsResponse(**response.json())
```

Comment:
Should use `model_validate`. Also have just one call where `job_id` is optional.

Reply:
Where job_id is optional?